cliffordlayers/models/gcan_layers.py

```diff
@@ -0,0 +1,349 @@
+import math
+import torch
+from torch import nn
+from torch.nn.modules.utils import _pair
+from cliffordlayers.nn.modules.cliffordg3conv import clifford_g3convnd
+
+
+def get_clifford_left_kernel(M, w, flatten=True):
+    o, i, c = w.size()
+    k = torch.einsum(f"ijk, pqi->jpkq", M, w)
+    if flatten:
+        k = k.reshape(o * c, i * c)
+    return k
+
+
+def get_clifford_right_kernel(M, w, flatten=True):
+    o, i, c = w.size()
+    k = torch.einsum(f"ijk, pqk->jpiq", M, w)
+    if flatten:
+        k = k.reshape(o * c, i * c)
+    return k
+
+
+class PGAConjugateLinear(nn.Module):
+    def __init__(
+        self,
+        in_features,
+        out_features,
+        algebra,
+        input_blades,
+        action_blades=(0, 5, 6, 7, 8, 9, 10, 15),
+    ):
+        super().__init__()
+        assert torch.all(algebra.metric == torch.tensor([0, 1, 1, 1]))
+        self.input_blades = input_blades
+        self.in_features = in_features
+        self.out_features = out_features
+        self.algebra = algebra
+        self.action_blades = action_blades
+        self.n_action_blades = len(action_blades)
+        self._action = nn.Parameter(
+            torch.empty(out_features, in_features, self.n_action_blades)
+        )
+        self._weight = nn.Parameter(torch.empty(out_features, in_features))
+        self._embed_e0 = nn.Parameter(torch.zeros(in_features, 1))
+
+        self.inverse = algebra.reverse
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        # Init the rotation parts uniformly.
+        torch.nn.init.uniform_(self._action[..., 0], -1, 1)
+        torch.nn.init.uniform_(self._action[..., 4:7], -1, 1)
+
+        # Init the translation parts with zeros.
+        torch.nn.init.zeros_(self._action[..., 1:4])
+        torch.nn.init.zeros_(self._action[..., 7])
+
+        norm = self.algebra.norm(
+            self.algebra.embed(self._action.data, self.action_blades)
+        )
+        assert torch.allclose(norm[..., 1:], torch.tensor(0.0), atol=1e-3)
+        norm = norm[..., :1]
+        self._action.data = self._action.data / norm
+
+        torch.nn.init.kaiming_uniform_(self._weight, a=math.sqrt(5))
+
+    @property
+    def action(self):
+        return self.algebra.embed(self._action, self.action_blades)
+
+    @property
+    def weight(self):
+        return self._weight
```

Comment:
Is it necessary to hide it this way? Why not just do `self.weight` above?

Reply:
Updated as you suggested.